Tests/SourceControlTests/GitRepositoryTests.swift

```diff
@@ -356,9 +356,21 @@ class GitRepositoryTests: XCTestCase {
 
             // We should have commits which are not pushed.
             XCTAssert(try checkoutRepo.hasUnpushedCommits())
-            // Push the changes and check again.
+            
+            // Detach the current HEAD
+            try systemQuietly([Git.tool, "-C", checkoutPath.asString, "checkout", "--detach"])
+            
+            // The detached HEAD should still not be pushed.
+            XCTAssert(try checkoutRepo.hasUnpushedCommits())
+            
+            // Restore the branch, push the changes and check again.
+            try checkoutTestRepo.checkout(revision: Revision(identifier: "master"))
             try checkoutTestRepo.push(remote: "origin", branch: "master")
             XCTAssertFalse(try checkoutRepo.hasUnpushedCommits())
+            
+            // Detach again, the changes should still be pushed.
+            try systemQuietly([Git.tool, "-C", checkoutPath.asString, "checkout", "--detach"])
+            XCTAssertFalse(try checkoutRepo.hasUnpushedCommits())
```

Comment:
I think the bug is for detecting the changes made on a detached head. I would leave the original test as is and add more lines to test that. Something along the lines of:

```
git checkout some-detached-head
touch a
git add . && git commit -m update
...assert we can detect the commit we just made...

git checkout master
...assert we can detect the commit we just made...

..push the commit..
...assert there are no more unpushed commits...
```